src/include/duckdb/common/cycle_counter.hpp

```diff
@@ -10,32 +10,72 @@
 
 #include "duckdb/common/helper.hpp"
 #include "duckdb/common/chrono.hpp"
+#include "duckdb/common/random_engine.hpp"
 
 namespace duckdb {
 
-//! The cycle counter can be used to measure elapsed cycles
+//! The cycle counter can be used to measure elapsed cycles for a function, expression and ...
+//! Optimized by sampling mechanism. Once per 100 times.
+//! //Todo Can be optimized further by calling RDTSC once per sample
 class CycleCounter {
 public:
-	//! Starts the timer
-	void Start() {
-		finished = false;
-		start = Tick();
+	CycleCounter() : random(0) {
 	}
-	//! Finishes timing
-	void End() {
-		end = Tick();
-		finished = true;
+	// Next_sample determines if a sample needs to be taken, if so start the profiler
+	void TakeSample() {
+		if (current_count >= next_sample) {
+			tmp = Tick();
+		}
 	}
 
-	uint64_t Elapsed() const {
-		return end - start;
+	// Process the sample
+	void ProcessSample(int chunk_size) {
+		if (current_count >= next_sample) {
+			time += Tick() - tmp;
+		}
+		if (current_count >= next_sample) {
+			next_sample = 50 + random.NextRandomInteger() % 100;
```

Comment:
For readability, please take the 50 and 100 out and put them in some static constexpr's (called e.g. sampling rate, sampling variance)